README.md
<h1 align="center" style="font-family:'Courier New'">Hebrew("בְּרֵאשִׁ֖ית")</h1>
<p align="center">
    <em>A python package with methods to handle the complexities of Hebrew text, calculate Gematria, and more.</em>
</p>
<p align="center">
<a href="https://pypi.org/project/hebrew/" target="_blank">
    <img src="https://badge.fury.io/py/hebrew.svg" alt="PyPI version">
</a>
<a href="https://badge.fury.io/py/hebrew">
    <img src="https://img.shields.io/pypi/pyversions/hebrew" alt="Supported Python Versions">
</a>
<a href="https://github.com/avi-perl/hebrew/actions/workflows/test.yml" target="_blank">
    <img src="https://github.com/avi-perl/hebrew/actions/workflows/test.yml/badge.svg" alt="Test">
</a>
<a href="https://codecov.io/gh/avi-perl/hebrew" target="_blank">
  <img src="https://codecov.io/gh/avi-perl/Hebrew/branch/master/graph/badge.svg?token=0DA1R9IY6S"/>
</a>
󠀠󠀠
<a href="https://twitter.com/__aviperl__" target="_blank">
    <img src="https://badgen.net/badge/icon/twitter?icon=twitter&label=Chat%20with%20me" alt="Twitter">
</a>
</p>

---

**Documentation**: [https://hebrew.aviperl.me/](https://hebrew.aviperl.me/)

**Repository**: [https://github.com/avi-perl/hebrew](https://github.com/avi-perl/hebrew)

---

# Installation

<!--pytest.mark.skip-->

```bash
$ pip install hebrew
```

# Example

`Hebrew` assists in working with Hebrew text by providing methods to handle the text according to user-perceived
characteristics. Additionally, methods for common Hebrew text processing are provided.

```python
from hebrew import Hebrew
from hebrew.chars import HebrewChar, ALEPH

hs = Hebrew('בְּרֵאשִׁ֖ית')
print(list(hs.graphemes))  # ['בְּ', 'רֵ', 'א', 'שִׁ֖', 'י', 'ת']
print(hs.text_only())  # בראשית

print(ALEPH)  # HebrewChar(char='א', name='Aleph', hebrew_name='אָלֶף', name_alts=['Alef'], hebrew_name_alts=None, final_letter=False)

print(HebrewChar.search('bet'))  # HebrewChar(char='בּ', name='Bet', hebrew_name='בֵּית', name_alts=None, hebrew_name_alts=None, final_letter=False)
```

## Gematria

The `Hebrew` class includes a `gematria` function that can return a value for _23_ different variations of Gematria!

```python
from hebrew import Hebrew
from hebrew import GematriaTypes

hs = Hebrew('בְּרֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃')

print(hs.gematria())  # 2701
print(hs.gematria(GematriaTypes.MISPAR_GADOL))  # 4631
```

Messy inputs, such as strings with english text mixed in, is supported. However, do be careful to work with sanitized
strings as much as possible.

```python
from hebrew import Hebrew

hs1 = Hebrew(
    '''
    Text:          "בְּרֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃"
    Translation:   "When God began to create heaven and earth"
    '''
)
hs2 = Hebrew('בְּרֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃')

assert hs1.gematria() ==  hs2.gematria()  # 2701
```

Major kudos goes to [TorahCalc](https://www.torahcalc.com/gematria/) whose calculator and explanations were critical to
the development of this feature.

## Numerical Conversions

You can create a Hebrew object that represents a number using the `from_number` method. This is handy for displaying numbers in Hebrew form.

```python
from hebrew import Hebrew

hs1 = Hebrew.from_number(2)
print(hs1)  # ב׳

# Add an apostrophe instead of a the unicode geresh
hs2 = Hebrew.from_number(2, geresh=False)
print(hs2)  # 'ב

# Do not add punctuation
hs2 = Hebrew.from_number(2, geresh=False)
print(hs2)  # ב

```

## Grapheme Characters

Hebrew text comes in different forms, depending on the context. Hebrew text may appear with Niqqudot
"a system of diacritical signs used to represent vowels or distinguish between alternative pronunciations of letters
of the Hebrew alphabet". [^1] Additionally, Hebrew text may appear with extensive punctuation characters that connect
words, separate them, and cantillation marks "used as a guide for chanting the text, either from the printed text or,
in the case of the public reading of the Torah" [^2].

Because of the above, from the perspective of a hebrew reader, the following 3 words are the same:

1. **בְּרֵאשִׁ֖ית**
2. **בְּרֵאשִׁית**
3. **בראשית**

However, as a unicode string, they are entirely different because of the additional characters.

```python
assert len("בְּרֵאשִׁ֖ית") == 12
assert len("בְּרֵאשִׁית") == 11
assert len("בראשית") == 6
```

This impacts the user is a number of other ways. For example, if I want to get the root of this hebrew word using a slice:
_Expected: `רֵאשִׁ֖ית`_

```python
he = "בְּרֵאשִׁ֖ית"
assert he[-5:] == 'ִׁ֖ית'
```

The solution to this is to handle the unicode string as a list of grapheme[^3] characters, where each letter and its
accompanying characters are treated as a single unit.

### Working with Grapheme Characters

Using the [grapheme](https://github.com/alvinlindstam/grapheme) library for python, we can work with the grapheme
characters as units. This allows us to get the right number of characters, slice the string correctly, and more.

```python
import grapheme

assert grapheme.length("בְּרֵאשִׁ֖ית") == 6
assert grapheme.slice("בְּרֵאשִׁ֖ית", start=1, end=6) == 'רֵאשִׁ֖ית'
```

This library includes 2 classes. `GraphemeString` is a class that supports all the functions made available by `grapheme`.
The 2nd class `Hebrew` subclasses `GraphemeString` and adds methods for handling Hebrew text. This allows us to
interact with the text like so:

```python
from hebrew import Hebrew

v2 = Hebrew("וְהָאָ֗רֶץ הָיְתָ֥ה תֹ֙הוּ֙ וָבֹ֔הוּ וְחֹ֖שֶׁךְ עַל־פְּנֵ֣י תְה֑וֹם וְר֣וּחַ אֱלֹהִ֔ים מְרַחֶ֖פֶת עַל־פְּנֵ֥י הַמָּֽיִם׃")
print(v2.no_taamim())  # "וְהָאָרֶץ הָיְתָה תֹהוּ וָבֹהוּ וְחֹשֶׁךְ עַל־פְּנֵי תְהוֹם וְרוּחַ אֱלֹהִים מְרַחֶפֶת עַל־פְּנֵי הַמָּיִם׃"
print(v2.text_only())  # והארץ היתה תהו ובהו וחשך על־פני תהום ורוח אלהים מרחפת על־פני המים

assert v2.length == 66
print(v2.words(split_maqaf=True))  # [וְהָאָ֗רֶץ, הָיְתָ֥ה, תֹ֙הוּ֙, וָבֹ֔הוּ, וְחֹ֖שֶׁךְ, עַל, פְּנֵ֣י, תְה֑וֹם, וְר֣וּחַ, אֱלֹהִ֔ים, מְרַחֶ֖פֶת, עַל, פְּנֵ֥י, הַמָּֽיִם׃]
```

The text in these examples and used in testing were sourced from [Sefaria](https://github.com/Sefaria/Sefaria-Export).

## `hebrew.chars` and Character Constants

`hebrew.Chars` contains constants for every letter as well as lists by character category's.
Each value is an instance of a class that represents a character in the Hebrew character set with relevant properties.
Since this library seeks to support the use of the Hebrew language in the way it is used, characters such as "בּ" can be
located (`BET`) even though, strictly speaking, "בּ" is not part of the hebrew alphabet; it is a Hebrew letter plus a dot.

```python
from hebrew.chars import FINAL_LETTERS, YIDDISH_CHARS, TSADI

print(TSADI)  # HebrewChar(char='צ', name='Tsadi', hebrew_name='צַדִי', name_alts=['Tzadik'], hebrew_name_alts=['צדיק'], final_letter=False)

assert {c.name: c.char for c in FINAL_LETTERS} == {'Chaf Sofit': 'ך', 'Mem Sofit': 'ם', 'Nun Sofit': 'ן', 'Fe Sofit': 'ף', 'Tsadi Sofit': 'ץ'}

assert [c.char for c in YIDDISH_CHARS] == ['ײ', 'װ', 'ױ']
```

A letter can be retrieved using the `CHARS` dict; A dict of all instances of all supported Char types where the key is
the char and the value is an instance of BaseHebrewChar.

```python
from hebrew.chars import CHARS

print(CHARS.get('בּ'))  # HebrewChar(char='בּ', name='Bet', hebrew_name='בֵּית', name_alts=None, hebrew_name_alts=None, final_letter=False)
```

Search is also supported so that letters can be retrieved by their name.

```python
from hebrew.chars import HebrewChar

print(HebrewChar.search('bet'))  # HebrewChar(char='בּ', name='Bet', hebrew_name='בֵּית', name_alts=None, hebrew_name_alts=None, final_letter=False)
```

## Contributing

Contributions in the form of pull requests are very welcome! I'm sure many more helpful methods related to hebrew text
could be helpful. More information and instructions for contributing can be found [here](CONTRIBUTING).

[^1]: [https://en.wikipedia.org/wiki/Niqqud](https://en.wikipedia.org/wiki/Niqqud)
[^2]: [https://en.wikipedia.org/wiki/Hebrew_cantillation](https://en.wikipedia.org/wiki/Hebrew_cantillation)
[^3]: [https://en.wikipedia.org/wiki/Grapheme](https://en.wikipedia.org/wiki/Grapheme)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.8.1] - 2023-12-11

### Fixed

- Corrected an issue with subclasses of `BaseHebrewChar` where use of `__eq__` would break when the type for the 
  comparison was not the same as itself.

## [0.8.0] - 2023-12-10

### Added

- Added `Hebrew.normalize`, a function for normalizing the hebrew characters in a string. This is typically needed with 
text includes special hebrew characters.

  Hidden among hebrew text can be special characters that are visually identical humans, but are made up of different 
  unicode characters. However, this can cause issues with presentation when there is no support for these characters.

  ![bad_character_display.png](docs/img/bad_character_display.png)

  In this case, the first letter is made up of 2 unicode characters, [sin with a dot](https://en.wiktionary.org/wiki/%D7%A9%D7%82)
  and [qamatz](https://en.wiktionary.org/wiki/%D6%B8). The issue here is the sin. By normalizing the sin with a dot to 2 unicode 
  characters, [ש](https://en.wiktionary.org/wiki/%D7%A9) and the [dot](https://en.wiktionary.org/wiki/%D7%82), the display 
  will look right!

  ![good_character_display.png](docs/img/good_character_display.png)

  To normalize content, use the `Hebrew.normalize` function:
  ```python
  from hebrew import Hebrew

  hs = Hebrew('שָׂחַקְתִּי כְּמוֹ')

  assert len(hs.string) == 14
  assert len(hs.normalize().string) == 18
  ```

  ### Normalizing Yiddish
  
  By default, special yiddish characters such as [ײ](https://en.wiktionary.org/wiki/%D7%B2) (double Yod) are _not_ normalized.
  However, [ײַ](https://en.wiktionary.org/wiki/%EF%AC%9F) (double Yod with a Patah) will be converted to [ײַ](https://en.wiktionary.org/wiki/%D7%B2%D6%B7).
  
  To fully "normalize" yiddish characters, pass `True` to `normalize`.


## [0.7.0] - 2023-12-10

### Added

- New function `Hebrew.from_number` converts an int into its hebrew form

### Example:

``` python
from hebrew import Hebrew

hs1 = Hebrew.from_number(2)
print(hs1)  # ב׳

# Do not add punctuation
hs2 = Hebrew.from_number(2, geresh=False)
print(hs2)  # ב
```

## [0.6.2] - 2023-12-08

### Changed

- Added 3.12 as a supported version
- Updated project dependencies.

### Removed
- Removed 3.11 as a supported version

## [0.6.1] - 2022-10-28

### Changed

- Added 3.11 as a supported version

## [0.6.0] - 2022-04-19

### Added

- Added the remaining missing Gematria methods. The complete list of supported methods is: 
  - `ACHAS_BETA`
  - `ALBAM`
  - `ATBASH`
  - `AVGAD`
  - `AYAK_BACHAR`
  - `MISPAR_BONEEH`
  - `MISPAR_GADOL`
  - `MISPAR_HAACHOR`
  - `MISPAR_HAMERUBAH_HAKLALI`
  - `MISPAR_HECHRACHI`
  - `MISPAR_KATAN`
  - `MISPAR_KATAN_MISPARI`
  - `MISPAR_KIDMI`
  - `MISPAR_KOLEL`
  - `MISPAR_MESHULASH`
  - `MISPAR_MISPARI`
  - `MISPAR_MUSAFI`
  - `MISPAR_NEELAM`
  - `MISPAR_PERATI`
  - `MISPAR_SHEMI_MILUI`
  - `MISPAR_SIDURI`
  - `OFANIM`
  - `REVERSE_AVGAD`
  
  ### Example:
  ```python
  from hebrew import Hebrew
  from hebrew import GematriaTypes
  
  hs = Hebrew('בְּרֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָֽרֶץ׃')
  
  print(hs.gematria())  # 2701
  print(hs.gematria(GematriaTypes.MISPAR_GADOL))  # 4631
  ```

## [0.5.8] - 2022-04-18

### Added

- Added Gematria methods `MISPAR_MUSAFI`. This is the first of the complex Gematria types that goes deeper than just 
adding up values assigned to each letter, necessitating internal changes to the `Hebrew` class.

## [0.5.7] - 2022-04-18

### Added

- Added Gematria methods `MISPAR_KATAN`, `MISPAR_PERATI`, `ATBASH`, `ALBAM`, `MISPAR_MESHULASH`.

## [0.5.6] - 2022-04-18

### Added

- Added a new gematria method, `MISPAR_SIDURI`.

## [0.5.5] - 2021-11-22

### Added

- Added a new gematria method, `MISPAR_GADOL`. A contribution by [Taber Andrew Bain](https://github.com/taber)

## [0.5.4] - 2021-11-21

### Fixed

- Fixed an issue where `Hebrew.gematria` would through an error if the input string had no hebrew characters.
  In this case, we now return a value of 0.

## [0.5.3] - 2021-11-15

### Changed

- Split the `PunctuationChar` type chars into `TaamimChar` and `OtherChar` types in `hebrew.char`.
- Renamed the `no_punctuation` method of `Hebrew` to `no_taamim`.

## [0.5.0] - 2021-11-14

### Added

- Added the method `Hebrew.gematria` method for calculating the gematria of a string.
- Added `mispar_hechrachi` as a supported gematria type.

``` python
>>> from hebrew import Hebrew
>>> from hebrew.gematria import GematriaTypes

>>> Hebrew("בְּרֵאשִׁ֖ית בָּרָ֣א אֱלֹהִ֑ים אֵ֥ת הַשָּׁמַ֖יִם וְאֵ֥ת הָאָרֶץ׃").gematria(GematriaTypes.MISPAR_HECHRACHI)
2701
```

## [0.4.0] - 2021-11-14

### Added

- Added `hebrew.chars` with constants for Hebrew characters and classes to represent each letter.
- Moved constants out of `Hebrew` and into `hebrew.chars`.
- Constant values, previously strings, are now instances of a class with metadata for each letter.

### Removed

- Support for Python 3.6 was removed because we are now using `@dataclasse`. It is possible to make this work with
  3.6 but I am choosing not to at this time. If this is a problem for you, feel free to open an issue.

## [0.3.0] - 2021-11-08

### Changed

- Renamed the python package from `hebrewstring` to `hebrew`.

## [0.2.0] - 2021-11-07

### Added

- Added the `__eq__` method to the `GraphemeString` object.

  This is to support the `==` operator when comparing two `GraphemeString` objects.

- Added the `__add__` method to the `GraphemeString` object.

  This is to support the `+` operator when adding two `GraphemeString` objects together.

- Added the `__hash__` method to the `GraphemeString` object.

  This is to support the `hash()` function for a `GraphemeString` instance and allows it (as an example) to be used as
  a `dict` key.


hebrew/gematria.py
from enum import Enum
from typing import Dict


class GematriaTypes(Enum):
    """
    Types of Gematria supported in this library.
    """

    # Simple methods where the value is calculated off of a value map to each letter
    MISPAR_HECHRACHI = "mispar_hechrachi"
    MISPAR_GADOL = "mispar_gadol"
    MISPAR_SIDURI = "mispar_siduri"
    MISPAR_KATAN = "mispar_katan"
    MISPAR_PERATI = "mispar_perati"
    ATBASH = "atbash"
    ALBAM = "albam"
    MISPAR_MESHULASH = "mispar_meshulash"
    MISPAR_KIDMI = "mispar_kidmi"
    MISPAR_MISPARI = "mispar_mispari"
    AYAK_BACHAR = "ayak_bachar"
    AYAK_BAKAR: str = AYAK_BACHAR  # type: ignore
    OFANIM = "ofanim"
    ACHAS_BETA = "achas_beta"
    AVGAD = "avgad"
    REVERSE_AVGAD = "reverse_avgad"

    # Complex methods involving logic on top of MISPAR_HECHRACHI
    MISPAR_MUSAFI = "mispar_musafi"
    MISPAR_BONEEH = "mispar_boneeh"
    MISPAR_HAMERUBAH_HAKLALI = "mispar_hamerubah_haklali"
    MISPAR_HAACHOR = "mispar_haachor"
    MISPAR_KATAN_MISPARI = "mispar_katan_mispari"
    MISPAR_KOLEL = "mispar_kolel"
    MISPAR_SHEMI_MILUI = "mispar_shemi_milui"
    MISPAR_NEELAM = "mispar_neelam"


"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the mispar_hechrachi
gematria method as its value.
"""
MISPAR_HECHRACHI: Dict[str, int] = {
    "א": 1,
    "ב": 2,
    "ג": 3,
    "ד": 4,
    "ה": 5,
    "ו": 6,
    "ז": 7,
    "ח": 8,
    "ט": 9,
    "י": 10,
    "כ": 20,
    "ך": 20,
    "ל": 30,
    "מ": 40,
    "ם": 40,
    "נ": 50,
    "ן": 50,
    "ס": 60,
    "ע": 70,
    "פ": 80,
    "ף": 80,
    "צ": 90,
    "ץ": 90,
    "ק": 100,
    "ר": 200,
    "ש": 300,
    "ת": 400,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the mispar_gadol
gematria method as its value. Non-final letter values are all the same, but final kaf/chaf, mem, nun, pe/fe are
moved to the end of the alphabet and continue on from where tav left off
"""
MISPAR_GADOL: Dict[str, int] = {
    "א": 1,
    "ב": 2,
    "ג": 3,
    "ד": 4,
    "ה": 5,
    "ו": 6,
    "ז": 7,
    "ח": 8,
    "ט": 9,
    "י": 10,
    "כ": 20,
    "ך": 500,
    "ל": 30,
    "מ": 40,
    "ם": 600,
    "נ": 50,
    "ן": 700,
    "ס": 60,
    "ע": 70,
    "פ": 80,
    "ף": 800,
    "צ": 90,
    "ץ": 900,
    "ק": 100,
    "ר": 200,
    "ש": 300,
    "ת": 400,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the mispar_siduri
gematria method as its value. This method assigns each letter a number from 1 to 22 in the order of the alphabet.
"""
MISPAR_SIDURI: Dict[str, int] = {
    "א": 1,
    "ב": 2,
    "ג": 3,
    "ד": 4,
    "ה": 5,
    "ו": 6,
    "ז": 7,
    "ח": 8,
    "ט": 9,
    "י": 10,
    "כ": 11,
    "ך": 23,
    "ל": 12,
    "מ": 13,
    "ם": 24,
    "נ": 14,
    "ן": 25,
    "ס": 15,
    "ע": 16,
    "פ": 17,
    "ף": 26,
    "צ": 18,
    "ץ": 27,
    "ק": 19,
    "ר": 20,
    "ש": 21,
    "ת": 22,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the mispar_katan
gematria method as its value. This method uses the value of the letters but without the zeros after large numbers. 
(ex. "Yud" is 1 instead of 10, "Tav" is 4 instead of 400).
"""
MISPAR_KATAN: Dict[str, int] = {
    "א": 1,
    "ב": 2,
    "ג": 3,
    "ד": 4,
    "ה": 5,
    "ו": 6,
    "ז": 7,
    "ח": 8,
    "ט": 9,
    "י": 1,
    "כ": 2,
    "ך": 2,
    "ל": 3,
    "מ": 4,
    "ם": 4,
    "נ": 5,
    "ן": 5,
    "ס": 6,
    "ע": 7,
    "פ": 8,
    "ף": 8,
    "צ": 9,
    "ץ": 9,
    "ק": 1,
    "ר": 2,
    "ש": 3,
    "ת": 4,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the mispar_perati
gematria method as its value. This method assigns each letter its standard value squared. (ex. "Aleph" = 1 x 1 = 1, "Beis" = 2 x 2 = 4).
"""
MISPAR_PERATI: Dict[str, int] = {
    "א": 1,
    "ב": 4,
    "ג": 9,
    "ד": 16,
    "ה": 25,
    "ו": 36,
    "ז": 49,
    "ח": 64,
    "ט": 81,
    "י": 100,
    "כ": 400,
    "ך": 400,
    "ל": 900,
    "מ": 1_600,
    "ם": 1_600,
    "נ": 2_500,
    "ן": 2_500,
    "ס": 3_600,
    "ע": 4_900,
    "פ": 6_400,
    "ף": 6_400,
    "צ": 8_100,
    "ץ": 8_100,
    "ק": 10_000,
    "ר": 40_000,
    "ש": 90_000,
    "ת": 160_000,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the AtBash
gematria method as its value. This method exchanges each letter's value for its opposite letter's value. 
(ex. "Aleph" switches values with "Tav", "Beis" switches values with "Shin").
"""
ATBASH: Dict[str, int] = {
    "א": 400,
    "ב": 300,
    "ג": 200,
    "ד": 100,
    "ה": 90,
    "ו": 80,
    "ז": 70,
    "ח": 60,
    "ט": 50,
    "י": 40,
    "כ": 30,
    "ך": 30,
    "ל": 20,
    "מ": 10,
    "ם": 10,
    "נ": 9,
    "ן": 9,
    "ס": 8,
    "ע": 7,
    "פ": 6,
    "ף": 6,
    "צ": 5,
    "ץ": 5,
    "ק": 4,
    "ר": 3,
    "ש": 2,
    "ת": 1,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the AlBam
gematria method as its value. This method splits the alphabet in half and letters from the first half switch 
values with letters from the second half. (ex. "Aleph" switches values with "Lamed", "Beis" switches values with "Mem").
"""
ALBAM: Dict[str, int] = {
    "א": 30,
    "ב": 40,
    "ג": 50,
    "ד": 60,
    "ה": 70,
    "ו": 80,
    "ז": 90,
    "ח": 100,
    "ט": 200,
    "י": 300,
    "כ": 400,
    "ך": 400,
    "ל": 1,
    "מ": 2,
    "ם": 2,
    "נ": 3,
    "ן": 3,
    "ס": 4,
    "ע": 5,
    "פ": 6,
    "ף": 6,
    "צ": 7,
    "ץ": 7,
    "ק": 8,
    "ר": 9,
    "ש": 10,
    "ת": 20,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the AtBash
gematria method as its value. This method splits the alphabet in half and letters from the first half switch 
values with letters from the second half. (ex. "Aleph" switches values with "Lamed", "Beis" switches values with "Mem").
"""
MISPAR_MESHULASH: Dict[str, int] = {
    "א": 1,
    "ב": 8,
    "ג": 27,
    "ד": 64,
    "ה": 125,
    "ו": 216,
    "ז": 343,
    "ח": 512,
    "ט": 729,
    "י": 1_000,
    "כ": 8_000,
    "ך": 8_000,
    "ל": 27_000,
    "מ": 64_000,
    "ם": 64_000,
    "נ": 125_000,
    "ן": 125_000,
    "ס": 216_000,
    "ע": 343_000,
    "פ": 512_000,
    "ף": 512_000,
    "צ": 729_000,
    "ץ": 729_000,
    "ק": 1_000_000,
    "ר": 8_000_000,
    "ש": 27_000_000,
    "ת": 64_000_000,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Mispar Kidmi
gematria method as its value. This method adds the value of all preceding letters in the alphabet to each 
letter's value. (ex. "Aleph" = 1, "Beis" = 1 + 2 = 3, "Gimmel" = 1 + 2 + 3 = 6).
"""
MISPAR_KIDMI: Dict[str, int] = {
    "א": 1,
    "ב": 3,
    "ג": 6,
    "ד": 10,
    "ה": 15,
    "ו": 21,
    "ז": 28,
    "ח": 36,
    "ט": 45,
    "י": 55,
    "כ": 75,
    "ך": 75,
    "ל": 105,
    "מ": 145,
    "ם": 145,
    "נ": 195,
    "ן": 195,
    "ס": 255,
    "ע": 325,
    "פ": 405,
    "ף": 405,
    "צ": 495,
    "ץ": 495,
    "ק": 595,
    "ר": 795,
    "ש": 1095,
    "ת": 1495,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Mispar Mispari
gematria method as its value. This method spells out the Hebrew name of each of the letter's standard values and adds 
up their values. (ex. "Aleph" = one (Echad) = 1 + 8 + 4 = 13).
"""
MISPAR_MISPARI: Dict[str, int] = {
    "א": 13,
    "ב": 760,
    "ג": 636,
    "ד": 273,
    "ה": 348,
    "ו": 600,
    "ז": 372,
    "ח": 401,
    "ט": 770,
    "י": 570,
    "כ": 620,
    "ך": 620,
    "ל": 686,
    "מ": 323,
    "ם": 323,
    "נ": 408,
    "ן": 408,
    "ס": 660,
    "ע": 422,
    "פ": 446,
    "ף": 446,
    "צ": 820,
    "ץ": 820,
    "ק": 46,
    "ר": 501,
    "ש": 1083,
    "ת": 720,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Ayak Bachar 
(or Ayak Bakar) gematria method as its value. This method splits the alphabet into three groups of nine with the final 
(sofit) letters at the end. The letters in the first group replace the ones in the second group, the letters in the 
second group replace the ones in the third group, and the letters in the third group replace the ones in 
the first group. (ex. "Aleph" takes the place of "Yud", "Yud" takes the place of "Kuf", 
"Kuf" takes the place of "Aleph", "Beis" takes the place of "Chaf" etc.).
"""
AYAK_BACHAR: Dict[str, int] = {
    "א": 10,
    "ב": 20,
    "ג": 30,
    "ד": 40,
    "ה": 50,
    "ו": 60,
    "ז": 70,
    "ח": 80,
    "ט": 90,
    "י": 100,
    "כ": 200,
    "ך": 5,
    "ל": 300,
    "מ": 400,
    "ם": 6,
    "נ": 500,
    "ן": 7,
    "ס": 600,
    "ע": 700,
    "פ": 800,
    "ף": 8,
    "צ": 900,
    "ץ": 9,
    "ק": 1,
    "ר": 2,
    "ש": 3,
    "ת": 4,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Ofanim 
gematria method as its value. This method replaces each letter with the last letter of its name. (ex. "Aleph" becomes 
"Fey", "Beis" becomes "Tav").
"""
OFANIM: Dict[str, int] = {
    "א": 80,
    "ב": 400,
    "ג": 30,
    "ד": 400,
    "ה": 1,
    "ו": 6,
    "ז": 50,
    "ח": 400,
    "ט": 400,
    "י": 4,
    "כ": 80,
    "ך": 80,
    "ל": 4,
    "מ": 40,
    "ם": 40,
    "נ": 50,
    "ן": 50,
    "ס": 20,
    "ע": 50,
    "פ": 1,
    "ף": 1,
    "צ": 10,
    "ץ": 10,
    "ק": 80,
    "ר": 300,
    "ש": 50,
    "ת": 6,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Ofanim 
gematria method as its value. This method splits the alphabet into groups of 7, 7, and 8 letters. 
The letters in the first group replace the ones in the second group, the letters in the second group replace 
the ones in the third group, and the letters in the third group replace the ones in the first group. The letter "Tav" 
does not change.
"""
ACHAS_BETA: Dict[str, int] = {
    "א": 8,
    "ב": 9,
    "ג": 10,
    "ד": 20,
    "ה": 30,
    "ו": 40,
    "ז": 50,
    "ח": 60,
    "ט": 70,
    "י": 80,
    "כ": 90,
    "ך": 90,
    "ל": 100,
    "מ": 200,
    "ם": 200,
    "נ": 300,
    "ן": 300,
    "ס": 1,
    "ע": 2,
    "פ": 3,
    "ף": 3,
    "צ": 4,
    "ץ": 4,
    "ק": 5,
    "ר": 6,
    "ש": 7,
    "ת": 400,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Avgad  
gematria method as its value. This method replaces each letter with the next one. (ex. "Aleph" becomes "Beis", "Beis" 
becomes "Gimmel", "Tav" becomes "Aleph").
"""
AVGAD: Dict[str, int] = {
    "א": 2,
    "ב": 3,
    "ג": 4,
    "ד": 5,
    "ה": 6,
    "ו": 7,
    "ז": 8,
    "ח": 9,
    "ט": 10,
    "י": 20,
    "כ": 30,
    "ך": 30,
    "ל": 40,
    "מ": 50,
    "ם": 50,
    "נ": 60,
    "ן": 60,
    "ס": 70,
    "ע": 80,
    "פ": 90,
    "ף": 90,
    "צ": 100,
    "ץ": 100,
    "ק": 200,
    "ר": 300,
    "ש": 400,
    "ת": 1,
}

"""
A dictionary of values with each letter of the alphabet as a key, and the numerical value used in the Reverse Avgad  
gematria method as its value. This method replaces each letter with the previous one. (ex. "Beis" becomes "Aleph", 
"Gimmel" becomes "Beis", "Aleph" becomes "Tav").
"""
REVERSE_AVGAD: Dict[str, int] = {
    "א": 400,
    "ב": 1,
    "ג": 2,
    "ד": 3,
    "ה": 4,
    "ו": 5,
    "ז": 6,
    "ח": 7,
    "ט": 8,
    "י": 9,
    "כ": 10,
    "ך": 10,
    "ל": 20,
    "מ": 30,
    "ם": 30,
    "נ": 40,
    "ן": 40,
    "ס": 50,
    "ע": 60,
    "פ": 70,
    "ף": 70,
    "צ": 80,
    "ץ": 80,
    "ק": 90,
    "ר": 100,
    "ש": 200,
    "ת": 300,
}
